<a href="https://colab.research.google.com/github/gtoubian/cce/blob/main/4_3_Regression_Workshop_Gabriel_Toubian.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Gabriel Toubian** - gtoubian@gmail.com

#NOTE:
Please insert a text cell at the top of the Notebook with your Full Name in Bold and your email that you use for both Journey and Github

Also, add to the title of the notebook, an underscore as well as your name.

Ex. '4.2_Regression_Workshop_Usama'


#Warmup

Import pandas, numpy, scipy.stats, statsmodels.api, LinearRegression from sklearn & ols from statsmodels.formula.api.

In [169]:
import pandas as pd
import numpy as np
import scipy.stats as sc
import statsmodels.api as sm
from statsmodels.formula.api import ols
from sklearn.linear_model import LinearRegression

In [170]:
import matplotlib.pyplot as plt
import seaborn as sns

#Boston Regression

Using statsmodels and the `boston` dataset, make a regression model to predict house prices. Don't forget to add a constant (intercept) term. Note that statsmodels can take a `pd.DataFrame` as an input for `X`.

Report the $R^2$ and coefficients on each feature.

In [171]:
from sklearn.datasets import load_boston
import pandas as pd

Without Intercept

In [172]:
y = load_boston().target
c= boston
sm.OLS(y, c).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                      y   R-squared (uncentered):                   0.959
Model:                            OLS   Adj. R-squared (uncentered):              0.958
Method:                 Least Squares   F-statistic:                              891.3
Date:                Thu, 25 Mar 2021   Prob (F-statistic):                        0.00
Time:                        19:50:09   Log-Likelihood:                         -1523.8
No. Observations:                 506   AIC:                                      3074.
Df Residuals:                     493   BIC:                                      3128.
Df Model:                          13                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
CRIM          -0.0929      0.034     -2.699      0.007      -0.161      -0.025
ZN             0.0487      0.014      3.382      0.001       0.020       0.077
INDUS         -0.0041      0.064     -0.063      0.950      -0.131       0.123
CHAS           2.8540      0.904      3.157      0.002       1.078       4.630
NOX           -2.8684      3.359     -0.854      0.394      -9.468       3.731
RM             5.9281      0.309     19.178      0.000       5.321       6.535
AGE           -0.0073      0.014     -0.526      0.599      -0.034       0.020
DIS           -0.9685      0.196     -4.951      0.000      -1.353      -0.584
RAD            0.1712      0.067      2.564      0.011       0.040       0.302
TAX           -0.0094      0.004     -2.395      0.017      -0.017      -0.002
PTRATIO       -0.3922      0.110     -3.570      0.000      -0.608      -0.176
B              0.0149      0.003      5.528      0.000       0.010       0.020
LSTAT         -0.4163      0.051     -8.197      0.000      -0.516      -0.317
==============================================================================
Omnibus:                      204.082   Durbin-Watson:                   0.999
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1374.225
Skew:                           1.609   Prob(JB):                    3.90e-299
Kurtosis:                      10.404   Cond. No.                     8.50e+03
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 8.5e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

With Intercept

In [175]:
boston = pd.DataFrame(load_boston().data,columns = load_boston().feature_names )
y = load_boston().target

X = sm.add_constant(boston)
sm.OLS(y, X).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.741
Model:                            OLS   Adj. R-squared:                  0.734
Method:                 Least Squares   F-statistic:                     108.1
Date:                Thu, 25 Mar 2021   Prob (F-statistic):          6.72e-135
Time:                        19:51:03   Log-Likelihood:                -1498.8
No. Observations:                 506   AIC:                             3026.
Df Residuals:                     492   BIC:                             3085.
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         36.4595      5.103      7.144      0.000      26.432      46.487
CRIM          -0.1080      0.033     -3.287      0.001      -0.173      -0.043
ZN             0.0464      0.014      3.382      0.001       0.019       0.073
INDUS          0.0206      0.061      0.334      0.738      -0.100       0.141
CHAS           2.6867      0.862      3.118      0.002       0.994       4.380
NOX          -17.7666      3.820     -4.651      0.000     -25.272     -10.262
RM             3.8099      0.418      9.116      0.000       2.989       4.631
AGE            0.0007      0.013      0.052      0.958      -0.025       0.027
DIS           -1.4756      0.199     -7.398      0.000      -1.867      -1.084
RAD            0.3060      0.066      4.613      0.000       0.176       0.436
TAX           -0.0123      0.004     -3.280      0.001      -0.020      -0.005
PTRATIO       -0.9527      0.131     -7.283      0.000      -1.210      -0.696
B              0.0093      0.003      3.467      0.001       0.004       0.015
LSTAT         -0.5248      0.051    -10.347      0.000      -0.624      -0.425
==============================================================================
Omnibus:                      178.041   Durbin-Watson:                   1.078
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              783.126
Skew:                           1.521   Prob(JB):                    8.84e-171
Kurtosis:                       8.281   Cond. No.                     1.51e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.51e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [176]:
X = X.drop(columns=['INDUS', 'AGE'])
sm.OLS(y, X).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.741
Model:                            OLS   Adj. R-squared:                  0.735
Method:                 Least Squares   F-statistic:                     128.2
Date:                Thu, 25 Mar 2021   Prob (F-statistic):          5.54e-137
Time:                        19:51:08   Log-Likelihood:                -1498.9
No. Observations:                 506   AIC:                             3022.
Df Residuals:                     494   BIC:                             3072.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         36.3411      5.067      7.171      0.000      26.385      46.298
CRIM          -0.1084      0.033     -3.307      0.001      -0.173      -0.044
ZN             0.0458      0.014      3.390      0.001       0.019       0.072
CHAS           2.7187      0.854      3.183      0.002       1.040       4.397
NOX          -17.3760      3.535     -4.915      0.000     -24.322     -10.430
RM             3.8016      0.406      9.356      0.000       3.003       4.600
DIS           -1.4927      0.186     -8.037      0.000      -1.858      -1.128
RAD            0.2996      0.063      4.726      0.000       0.175       0.424
TAX           -0.0118      0.003     -3.493      0.001      -0.018      -0.005
PTRATIO       -0.9465      0.129     -7.334      0.000      -1.200      -0.693
B              0.0093      0.003      3.475      0.001       0.004       0.015
LSTAT         -0.5226      0.047    -11.019      0.000      -0.616      -0.429
==============================================================================
Omnibus:                      178.430   Durbin-Watson:                   1.078
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              787.785
Skew:                           1.523   Prob(JB):                    8.60e-172
Kurtosis:                       8.300   Cond. No.                     1.47e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.47e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Use polynomial features to improve your regression model in `2.1`. You can use squared and cubic features. Try to find a model that minimizes the `AIC` or `BIC` of your output table.

In [177]:
A = np.array(X)
print(A)

[[1.0000e+00 6.3200e-03 1.8000e+01 ... 1.5300e+01 3.9690e+02 4.9800e+00]
 [1.0000e+00 2.7310e-02 0.0000e+00 ... 1.7800e+01 3.9690e+02 9.1400e+00]
 [1.0000e+00 2.7290e-02 0.0000e+00 ... 1.7800e+01 3.9283e+02 4.0300e+00]
 ...
 [1.0000e+00 6.0760e-02 0.0000e+00 ... 2.1000e+01 3.9690e+02 5.6400e+00]
 [1.0000e+00 1.0959e-01 0.0000e+00 ... 2.1000e+01 3.9345e+02 6.4800e+00]
 [1.0000e+00 4.7410e-02 0.0000e+00 ... 2.1000e+01 3.9690e+02 7.8800e+00]]


In [ ]:
x = sm.add_constant(A)
A = np.concatenate([x, (x[:,6] ** 3)[:, np.newaxis]], axis=1)
print(A)

[[1.00000000e+00 6.32000000e-03 1.80000000e+01 ... 3.96900000e+02
  4.98000000e+00 6.84179290e+01]
 [1.00000000e+00 2.73100000e-02 0.00000000e+00 ... 3.96900000e+02
  9.14000000e+00 1.22548701e+02]
 [1.00000000e+00 2.72900000e-02 0.00000000e+00 ... 3.92830000e+02
  4.03000000e+00 1.22548701e+02]
 ...
 [1.00000000e+00 6.07600000e-02 0.00000000e+00 ... 3.96900000e+02
  5.64000000e+00 1.01830369e+01]
 [1.00000000e+00 1.09590000e-01 0.00000000e+00 ... 3.93450000e+02
  6.48000000e+00 1.36330777e+01]
 [1.00000000e+00 4.74100000e-02 0.00000000e+00 ... 3.96900000e+02
  7.88000000e+00 1.57189376e+01]]


In [ ]:
A = pd.DataFrame(A)

In [ ]:
A.columns = ['const', 'CRIM', 'ZN', 'CHAS', 'NOX', 'RM', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'NEW']

In [ ]:
A

const     CRIM    ZN  CHAS  ...  PTRATIO       B  LSTAT         NEW
0      1.0  0.00632  18.0   0.0  ...     15.3  396.90   4.98   68.417929
1      1.0  0.02731   0.0   0.0  ...     17.8  396.90   9.14  122.548701
2      1.0  0.02729   0.0   0.0  ...     17.8  392.83   4.03  122.548701
3      1.0  0.03237   0.0   0.0  ...     18.7  394.63   2.94  222.787480
4      1.0  0.06905   0.0   0.0  ...     18.7  396.90   5.33  222.787480
..     ...      ...   ...   ...  ...      ...     ...    ...         ...
501    1.0  0.06263   0.0   0.0  ...     21.0  391.99   9.67   15.227175
502    1.0  0.04527   0.0   0.0  ...     21.0  396.90   9.08   11.969701
503    1.0  0.06076   0.0   0.0  ...     21.0  396.90   5.64   10.183037
504    1.0  0.10959   0.0   0.0  ...     21.0  393.45   6.48   13.633078
505    1.0  0.04741   0.0   0.0  ...     21.0  396.90   7.88   15.718938

[506 rows x 13 columns]

In [ ]:
sm.OLS(y, A).fit().summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.750
Model:                            OLS   Adj. R-squared:                  0.744
Method:                 Least Squares   F-statistic:                     123.3
Date:                Thu, 25 Mar 2021   Prob (F-statistic):          6.56e-140
Time:                        17:14:59   Log-Likelihood:                -1489.4
No. Observations:                 506   AIC:                             3005.
Df Residuals:                     493   BIC:                             3060.
Df Model:                          12                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         45.3890      5.397      8.411      0.000      34.786      55.992
CRIM          -0.1262      0.032     -3.887      0.000      -0.190      -0.062
ZN             0.0311      0.014      2.269      0.024       0.004       0.058
CHAS           2.6999      0.839      3.217      0.001       1.051       4.349
NOX          -23.5572      3.753     -6.276      0.000     -30.932     -16.183
RM             3.7789      0.399      9.466      0.000       2.995       4.563
DIS           -2.7115      0.335     -8.100      0.000      -3.369      -2.054
RAD            0.3357      0.063      5.342      0.000       0.212       0.459
TAX           -0.0138      0.003     -4.123      0.000      -0.020      -0.007
PTRATIO       -1.0246      0.128     -8.001      0.000      -1.276      -0.773
B              0.0092      0.003      3.517      0.000       0.004       0.014
LSTAT         -0.5278      0.047    -11.324      0.000      -0.619      -0.436
NEW            0.0121      0.003      4.343      0.000       0.007       0.018
==============================================================================
Omnibus:                      150.684   Durbin-Watson:                   1.107
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              576.550
Skew:                           1.313   Prob(JB):                    6.36e-126
Kurtosis:                       7.522   Cond. No.                     1.65e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.65e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Note AIC and BIC are lower now

Ajusted R-Squared is 0.744

#Regressive Tennis T-Test

Use a linear regression and statsmodels to run a t-test on whether Federer scores more points than his opponents in the `tennis.csv` dataset.


Also answer the following: should your regression include a constant term? Why or why not? How would it change the interpretation of your coefficient and p-value?

In [ ]:
from google.colab import files
data = files.upload()

Saving tennis.csv to tennis (4).csv


In [ ]:
import io
tennis = pd.read_csv(io.BytesIO(data['tennis.csv']))

In [ ]:
tennis = tennis.dropna()

In [ ]:
score = np.array(tennis['win'])
sc.ttest_1samp(score, 0.5)

Ttest_1sampResult(statistic=26.245156533243442, pvalue=2.6923903460915165e-116)

In [ ]:
def one_sample_one_tailed(sample_data, popmean, alpha=0.05, alternative='greater'):
    t, p = sc.ttest_1samp(sample_data, popmean)
    print ('t:',t)
    print ('p:',p)
    if alternative == 'greater' and (p/2 < alpha) and t > 0:
        print ('Reject Null Hypothesis for greater-than test')
    if alternative == 'less' and (p/2 < alpha) and t < 0:
        print ('Reject Null Hypothesis for less-than test')

In [ ]:
one_sample_one_tailed(score, 0.5)

t: 26.245156533243442
p: 2.6923903460915165e-116
Reject Null Hypothesis for greater-than test


p-value is way below 0.05. We reject the null hypothesis. Federer wins more than he loses, so he does score more points than his opponent.

In [ ]:
tennis1 = tennis
tennis1 = tennis1.drop(columns = ['win'])

In [ ]:
tennis1 = tennis1.dropna()

In [ ]:
tennis['win'] = tennis['win'].astype(int)

In [ ]:
tenniswin = tennis['win']

In [205]:
tennis1

year atp points  ...  player2 total points won player2 total points total
0     1998          1  ...                      62.0                      103.0
1     1998         59  ...                      41.0                      106.0
2     1998         59  ...                      69.0                      148.0
3     1998         59  ...                      70.0                      130.0
5     1998          1  ...                      73.0                      136.0
...    ...        ...  ...                       ...                        ...
1173  2012        720  ...                      68.0                      164.0
1174  2012        720  ...                     146.0                      276.0
1175  2012         90  ...                      31.0                       86.0
1176  2012         90  ...                      39.0                       99.0
1177  2012         90  ...                      81.0                      173.0

[1016 rows x 54 columns]

In [197]:
tennis1.drop(tennis1.columns[[1,2,3,4,5,9,10,12,13,14,15,17,18,19,44,45]], axis = 1, inplace = True)

In [200]:
print(tennis3)

None


In [204]:

e = tenniswin
d = tennis1

sm.OLS(e, d).fit().summary()

ValueError: ignored

In [ ]:
e

0       0
1       1
2       1
3       0
5       0
       ..
1173    1
1174    0
1175    1
1176    1
1177    1
Name: win, Length: 1016, dtype: int64

In [180]:
d.head()

year  ... player2 total points total
0  1998  ...                      103.0
1  1998  ...                      106.0
2  1998  ...                      148.0
3  1998  ...                      130.0
5  1998  ...                      136.0

[5 rows x 70 columns]

#College Admissions

Using the `college.csv` dataset, answer the following:

1. Do private schools see more admissions overall? Hint: use a binary explanatory variable for `Private`

2. Is the relation between `Top10perc` and `Top25perc` best fit using a model with only one variable, or one variable and a polynomial of degree 2? Is a constant term useful? How would you select for the best of these model specifications?

In [ ]:
from google.colab import files
data = files.upload()

Saving college.csv to college.csv


In [ ]:
college = pd.read_csv(io.BytesIO(data['college.csv']))

In [ ]:
college

Unnamed: 0 Private  ...  Expend  Grad.Rate
0      Abilene Christian University     Yes  ...    7041         60
1                Adelphi University     Yes  ...   10527         56
2                    Adrian College     Yes  ...    8735         54
3               Agnes Scott College     Yes  ...   19016         59
4         Alaska Pacific University     Yes  ...   10922         15
..                              ...     ...  ...     ...        ...
772         Worcester State College      No  ...    4469         40
773               Xavier University     Yes  ...    9189         83
774  Xavier University of Louisiana     Yes  ...    8323         49
775                 Yale University     Yes  ...   40386         99
776    York College of Pennsylvania     Yes  ...    4509         99

[777 rows x 19 columns]

# Auto Multiple Regression

Using the `auto.csv` dataset, perform a multiple linear regression with `mpg` as the response and all other variables except name in `auto` as the predictors. Comment on the output:

i. Is there a relationship between the predictors and the response?

ii. Which predictors appear to have a statistically significant relationship to the response?

iii. What does the coefficient for the year variable suggest?